### PR TITLE
stdlib/system: add sink and move 

### PR DESCRIPTION
**Pull request by cooldome**

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -952,7 +952,7 @@ proc `&`*(x, y: char): string {.
   ##
   ## .. code-block:: Nim
   ##   assert('a' & 'b' == "ab")
-proc `&`*(x, y: string): string {.
+proc `&`*(x, y: sink string): string {.
```
Comment (Araq) [inline lib/system.nim]: No sink here.

Comment (cooldome) [reply lib/system.nim]: Yep

**Comment by cooldome** [comment]
This is PR ready for review. 
I wanted to provide details what I had to do to make it work.
Consider example from system.nim:
```nim
proc `&`*[T](x, y: sink seq[T]): seq[T] {.noSideEffect.} =
  newSeq(result, x.len + y.len)
  for i in 0..x.len-1:
    result[i] = move(x[i])
  for i in 0..y.len-1:
    result[i+x.len] = move(y[i])
```
This  proc needs move() calls in newruntime to be efficient. Unfortunately, these calls break old runtime, because of zeroing out of memory. Adding move() calls to this proc is enough to make compiler fail bootstrapping.  

Two ways forward: force `copyToSink()` analysis from injectdestructors to be applied to all types not just for objects with destructors. This as I discovered has negative effect on old runtime performance. Another way,I had adopted eventually, is to change move() proc such that it zeros out memory only if it is type with destructor.

**Comment by Araq** [comment]
But these semantics for ``move`` are weirder than before. I would have picked this solution:

```nim

when defined(gcDestructors):
  proc `&`*[T](x, y: sink seq[T]): seq[T] {.noSideEffect.} =
    newSeq(result, x.len + y.len)
    for i in 0..x.len-1:
      result[i] = move(x[i])
    for i in 0..y.len-1:
      result[i+x.len] = move(y[i])
else:
  # old `&` implementation
```

(Until we fixed the compiler so that ``sink`` always works, regardless of the ``--gc:arc`` switch.)

We are also working on ``sink`` parameter inference which can do partially what you did in this PR.

**Comment by cooldome** [comment]
@Araq. Ok, I can do that

**Comment by cooldome** [comment]
Ready for review

File: lib/system.nim [hunk 2]
```diff
@@ -1525,48 +1525,96 @@ proc `@`*[T](a: openArray[T]): seq[T] =
...
+    ## * `add(var seq[T], T) <#add,seq[T][T],T>`_
+    ##
+    ## .. code-block:: Nim
+    ##   assert(@[1, 2, 3] & 4 == @[1, 2, 3, 4])
+    newSeq(result, x.len + 1)
+    for i in 0..x.len-1:
+      result[i] = move(x[i])
+    result[x.len] = move(y)
+
+  proc `&`*[T](x: T, y: sink seq[T]): seq[T] {.noSideEffect.} =
```
Comment (Araq) [inline lib/system.nim]: No ``x: sink T`` here?

File: lib/system.nim [hunk 3]
```diff
@@ -1525,48 +1525,96 @@ proc `@`*[T](a: openArray[T]): seq[T] =
...
+    ## Prepends the element x to the beginning of the sequence.
+    ##
+    ## Requires copying of the sequence.
+    ##
+    ## .. code-block:: Nim
+    ##   assert(1 & @[2, 3, 4] == @[1, 2, 3, 4])
+    newSeq(result, y.len + 1)
+    result[0] = x
```
Comment (Araq) [inline lib/system.nim]: No ``move(x)`` here?

File: lib/system.nim (hunk 2 above)
Comment (cooldome) [reply lib/system.nim]: ouch, corrected

File: lib/system.nim (hunk 3 above)
Comment (cooldome) [reply lib/system.nim]: ouch, corrected

**Comment by cooldome** [comment]
Ready for review

**Comment by Araq** [comment]
Hint: These might use ``.nodestroy`` for even more efficient code.

**Pull request merged**